src/signature.c

```diff
@@ -335,6 +351,7 @@ struct p11prov_mech {
     int der_ecdsa_algorithm_id_len;
     const unsigned char *der_digestinfo;
     int der_digestinfo_len;
+    const EVP_MD *(*hash_md)(void);
```

Comment:
Why do we need this? We already have p11prov_digest_get_name() to get the name to be used in EVP_MD_fetch() like w already do in mech_fallback_init() or p11prov_digest_util() ...

Reply:
I probably misunderstood the OpenSSL API. Now reading that the `EVP_sha1()` has performance impact, while I though it will just return const reference to the existing instance.
I wanted to avoid doing the `EVP_MD_fetch()` as it involves searching for the actual implementation with context, propq ...
Given that we really want just the algorithm name, I think this would be good to simplify. Will have a look what can be done.

Reply:
Oh, thats the reason:

> Prior to OpenSSL 3.0, functions such as EVP_sha256() which return a "const" object were used directly to indicate the algorithm to use in various function calls.

Looking through the API, we could store the ASN1 DER format, convert it to ASN1_OBJECT using d2i_ASN1_OBJECT function and then pass the object to the algorithm with X509_ALGOR_set0(). Its not straightforward either, but avoids fetching the MD that many times. What do you think?

I would really really like to avoid constructing the Algorithm Identifier structure with PSS_PARAMS by hand. Its hell of nested sequences and I do not completely understand each of the encoding there.

Reply:
Or we could go through the NID with nid having stored in the table, then calling `OBJ_nid2obj()` and `X509_ALGOR_set0()`, similarly as done in `ossl_X509_ALGOR_from_nid()`. That sounds quite reasonable from my point of view.

Reply:
I got rid of the EVP_MD now.

The problem is that really all of the fields have default value so they are all optional and we need them in two contexts -- the key restrictions (where we might want to omit some) + signature parameters (we need to have all of them and we might need to support whole matrix of each SHA for digest x SHA for MGF1 (even though it does not make sense as PKCS#11 support this -- not sure if there is way to configure this in openssl though). 

https://datatracker.ietf.org/doc/html/rfc8017#appendix-A.2.3

Reply:
After all that RFC says:
```
      it is RECOMMENDED
      that the underlying hash function be the same as the one
      identified by hashAlgorithm
```

Reply:
Then I can see if I can get them pregenerated somehow reasonably. It should simplify things a bit ...